### PR TITLE
SALTO-5414 support polling as a capability in the new infra

### PR DESCRIPTION
**Pull request by Uri-Bar**
SALTO-5414 support polling as a capability in the new infra
---

_Additional context for reviewer_
added the polling option for both deploy + fetch
completed the support in additional status codes for deploy

---
_Release Notes_: 
none

---
_User Notifications_: 
none

### CODE REVIEW

File: packages/adapter-components/src/fetch/request/pagination/pagination.ts [hunk 1]
```diff
@@ -79,6 +86,46 @@ export const traversePages = async <ClientOptions extends string>({
...
         }
       }
+      const pollingFunc = async (
+        args: ClientBaseParams,
+        pollingArgs: PollingArgs,
+      ): Promise<Response<ResponseValue | ResponseValue[]> | undefined> => {
+        const response = await singleClientCall(args)
+        if (pollingArgs.checkStatus(response)) return response
+        return undefined
+      }
```
Comment (shir-reifenberg) [inline packages/adapter-components/src/fetch/request/pagination/pagination.ts:96]: This is almost identical to the function in `requester.ts` (in deploy), do you think we can consolidate ?

**Review by shir-reifenberg — COMMENTED** [review]
Looks good! 😎  
didn't review all of it yet, but looks like there are some similarities in fetch and deploy and I wonder if we can consolidate those.
Also, in order to make the code cleaner, I think we can extract the polling function to a separate function and make it get a 'makeRequest' (or any other name) function as a parameter and just do the polling on it. WDYT?

File: packages/adapter-components/src/definitions/system/requests/types.ts [hunk 2]
```diff
@@ -33,6 +34,14 @@ export type RequestArgs = {
   body?: unknown
 }
 
+export type PollingArgs = {
+  interval: number
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/adapter-components/src/definitions/system/requests/types.ts:38]: I had the same naming problem with checkStatus. the name probably should be `pollingInterval` and I think it is redundant

File: packages/adapter-components/src/definitions/system/requests/types.ts [hunk 3]
```diff
@@ -33,6 +34,14 @@ export type RequestArgs = {
...
+  retries: number
+  checkStatus: (response: Response<ResponseValue | ResponseValue[]>) => boolean
+  // fieldToPoll: string
+  // expectedValues: string[]
```
[the remark this answers is not in the document]
Comment (Uri-Bar) [reply packages/adapter-components/src/definitions/system/requests/types.ts:42]: yep, will remove

File: packages/adapter-components/src/fetch/request/pagination/pagination.ts (hunk 1 above)
Comment (Uri-Bar) [reply packages/adapter-components/src/fetch/request/pagination/pagination.ts:96]: will create polling file under client folder

File: packages/adapter-components/src/deployment/deploy/requester.ts [hunk 4]
```diff
@@ -237,13 +237,31 @@ export const getRequester = <TOptions extends APIDefinitionsOptions>({
...
+    const { polling } = mergedEndpointDef
 
-    return client[finalEndpointIdentifier.method ?? 'get']({
+    const singleClientCall = async (args: ClientDataParams): Promise<Response<ResponseValue | ResponseValue[]>> => {
+      try {
+        return await client[finalEndpointIdentifier.method ?? 'get'](args)
+      } catch (e) {
+        const status = e.response?.status
+        if (additionalValidStatuses.includes(status)) {
+          log.debug('Suppressing %d error %o', status, e)
```
Comment (shir-reifenberg) [inline packages/adapter-components/src/deployment/deploy/requester.ts:250]: can you please include method and path here? (same for fetch request) 🙏

File: packages/adapter-components/src/fetch/request/pagination/pagination.ts [hunk 5]
```diff
@@ -62,24 +71,32 @@ export const traversePages = async <ClientOptions extends string>({
...
-            log.debug('Suppressing %d error %o', status, e)
+            log.debug('Suppressing %d  fetch error %o', status, e)
             return { data: {}, status }
           }
           throw e
         }
       }
 
+      const processPage: ClientRequest = async args => {
+        //  SALTO-5575 consider splitting the polling from the pagination
```
Comment (shir-reifenberg) [inline packages/adapter-components/src/fetch/request/pagination/pagination.ts:88]:
```suggestion
        // TODO SALTO-5575 consider splitting the polling from the pagination
```

File: packages/adapter-components/src/client/polling.ts [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+} = retry
+
+export const executeWithPolling = async <T>(
+  args: T,
+  polling: PollingArgs,
+  singleClientCall: (args: T) => Promise<Response<ResponseValue | ResponseValue[]>>,
+): Promise<Response<ResponseValue | ResponseValue[]>> => {
+  const pollingFunc = async (): Promise<Response<ResponseValue | ResponseValue[]> | undefined> => {
+    const response = await singleClientCall(args)
+    if (polling.checkStatus(response)) return response
```
Comment (shir-reifenberg) [inline packages/adapter-components/src/client/polling.ts:32]: nit 

```suggestion
    return polling.checkStatus(response)) ? response : undefined
```

File: packages/adapter-components/src/client/polling.ts [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    return undefined
+  }
+
+  const pollingResult = await withRetry(() => pollingFunc(), {
```
Comment (shir-reifenberg) [inline packages/adapter-components/src/client/polling.ts:36]: maybe worth logging here here we passed max retries or something

**Review by shir-reifenberg — APPROVED** [review]
Looks great 👑  mainly nits

File: packages/adapter-components/src/deployment/deploy/requester.ts (hunk 4 above)
Comment (Uri-Bar) [reply packages/adapter-components/src/deployment/deploy/requester.ts:250]: ack, also added in the pagination

**Pull request merged**